src/Storages/MergeTree/MergeTreeData.cpp

```diff
@@ -1305,14 +1305,17 @@ void MergeTreeData::clearEmptyParts()
     if (!getSettings()->remove_empty_parts)
         return;
 
+    /// There is no need to wait for drop and hold thread in background pool.
+    auto context_for_drop = Context::createCopy(getContext());
+    context_for_drop->setSetting("replication_alter_partitions_sync", Field(0));
     auto parts = getDataPartsVector();
     for (const auto & part : parts)
     {
         if (part->rows_count == 0)
         {
             ASTPtr literal = std::make_shared<ASTLiteral>(part->name);
             /// If another replica has already started drop, it's ok, no need to throw.
-            dropPartition(literal, /* detach = */ false, /*drop_part = */ true, getContext(), /* throw_if_noop = */ false);
+            dropPartition(literal, /* detach = */ false, /*drop_part = */ true, context_for_drop, /* throw_if_noop = */ false);
```

Comment:
Looks ugly. Why don't extract another method from `dropPartition` that will do the internal parts without the whole query-related logic?